File: leetcode-submissions/0496-next-greater-element-i/solution.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # Create a result list 
        # Create hashmap key will be number and value will be index
        # Iterate through the nums1
        # Get the index in nums2 using hashmap and if there exists right index and it is greater than current number add it to the reuslt, else add -1
        
        hashmap = {}
        result = []
        for i, n in enumerate(nums2):
            hashmap[n] = i
            
        for num in nums1:
            idx = hashmap[num]
            p1 = idx
            if idx == len(nums2) - 1:
                result.append(-1)
                continue

            while True:
                if p1 == len(nums2) - 1:
                    result.append(-1)
                    break
                if nums2[p1+1] > nums2[idx]:
                    result.append(nums2[p1+1])
                    break
                p1 += 1

            
        return result

File: leetcode-submissions/0496-next-greater-element-i/solution.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        //just store the a next greater for each element in nums2... 
        
        //stack, if the next number is smaller add to stack
        //if it is greater, keep on popping off of the stack and adding in a hashmap

        Stack<Integer> stack = new Stack<>();
        Map<Integer, Integer> map = new HashMap<>();

        for(int i = 0; i < nums2.length; i++) {
            while(!stack.empty() && nums2[i] > stack.peek()) {
                map.put(stack.pop(), nums2[i]);
            }
            stack.push(nums2[i]);
        }
        
        int[] ans = new int[nums1.length];
        for(int k = 0; k <nums1.length; k++) {
            ans[k] = map.getOrDefault(nums1[k], -1);
        }
        return ans;   
    }
}
